File: Object/NormalMap.cpp
#include "Object/NormalMap.hpp"

namespace Object {
    NormalMap::NormalMap(std::unique_ptr<Object::Texture<3>> texture, float magnitude)
    {
        const TextureBase::MipMap &map = texture->mipMap(0);
        std::vector<float> values(map.width() * map.height() * 2);
        for (int j = 0; j < map.height(); j++) {
            for (int i = 0; i < map.width(); i++) {
                float s = 0;
                float su = 0;
                float sv = 0;
                for (int c = 0; c < 3; c++) {
                    s += map.at(i, j, c) / 3.0f;
                    su += map.at((i == map.width() - 1) ? 0 : i + 1, j, c) / 3.0f;
                    sv += map.at(i, (j == map.height() - 1) ? 0 : j + 1, c) / 3.0f;
                }
                values[(j * map.width() + i) * 2 + 0] = -(su - s) * map.width() * magnitude;
                values[(j * map.width() + i) * 2 + 1] = -(sv - s) * map.height() * magnitude;
            }
        }

        mTexture = std::make_unique<Object::Texture<2>>(map.width(), map.height(), std::move(values));
        mTexture->generateMipMaps();
    }

    Math::Normal NormalMap::perturbNormal(const Math::Point2D &surfacePoint, const Math::Bivector2D &surfaceProjection, const Math::Normal &normal, const Math::Bivector &tangent) const
    {
        Texture<2>::Value value = mTexture->sample(surfacePoint, surfaceProjection);
        Math::Vector offset = tangent * Math::Vector2D(value.channels[0], value.channels[1]);
        return (normal + Math::Normal(offset)).normalize();
    }

    void NormalMap::writeProxy(NormalMapProxy &proxy, OpenCL::Allocator &clAllocator) const
    {
        mTexture->writeProxy(proxy.texture, clAllocator);
    }
}

File: src/object/normal_map.rs
use crate::geo;
use crate::object;

use geo::Bivec2;
use geo::Bivec3;
use geo::Normal3;
use geo::Point2;
use geo::Vec2;

use object::Texture;

pub struct NormalMap {
    texture : Texture<2>
}

impl NormalMap {
    pub fn new(texture : Texture<3>, magnitude : f32) -> NormalMap {
        let map = &texture.mipmaps[0];
        let mut values = Vec::with_capacity(map.width * map.height * 2);
        for j in 0..map.height {
            for i in 0..map.width {
                let mut s = 0.0;
                let mut su = 0.0;
                let mut sv = 0.0;
                for c in 0..3 {
                    s += map.at(i, j, c) / 3.0;
                    su += map.at(if i == map.width { 0 } else { i + 1 }, j, c) / 3.0;
                    sv += map.at(i, if j == map.height { 0 } else { j + 1 }, c) / 3.0;
                }

                values.push(-(su - s) * (map.width as f32) * magnitude);
                values.push(-(sv - s) * (map.height as f32) * magnitude);
            }
        }

        let mut ntexture = Texture::new(map.width, map.height, values);
        ntexture.generate_mipmaps();

        return NormalMap {texture: ntexture};
    }

    pub fn perturb_normal(&self, surface_point : Point2, surface_projection : Bivec2, nrm : Normal3, tangent : Bivec3) -> Normal3 {
        let value = self.texture.sample(surface_point, surface_projection);
        let offset = tangent * Vec2::new(value[0], value[1]);
        return (nrm + Normal3::from_vec3(offset)).normalize();
    }
}